hibernate-core/src/main/java/org/hibernate/mapping/SingleTableSubclass.java

```diff
@@ -40,4 +43,62 @@ public void validate(Metadata mapping) throws MappingException {
 		}
 		super.validate( mapping );
 	}
+
+	@Override
+	public void createConstraints(MetadataBuildingContext context) {
+		if ( !isAbstract() ) {
+			final Dialect dialect = context.getMetadataCollector().getDatabase().getDialect();
+			if ( dialect.supportsTableCheck() ) {
+				final List<Selectable> discriminator = getDiscriminator().getSelectables();
+				if ( discriminator.size() == 1
+						&& discriminator.get( 0 ) instanceof Column discriminatorColumn ) {
+					final StringBuilder check = new StringBuilder();
+					check.append( discriminatorColumn.getQuotedName( dialect ) );
+					if ( isDiscriminatorValueNull() ) {
+						check.append( " is not null" );
+					}
+					if ( isDiscriminatorValueNotNull() ) {
+						check.append( " is null" );
+						// can't enforce this for now, because 'not null'
+						// really means "not null and not any of the other
+						// explicitly listed values"
+						return;  //ABORT
+					}
+					else {
+						check.append( " <> " );
+						final boolean quote = discriminatorColumn.getType().getJdbcType().isString();
+						if ( quote ) {
+							check.append( "'" );
+						}
+						check.append( getDiscriminatorValue() );
+						if ( quote ) {
+							check.append( "'" );
+						}
```

Comment:
Maybe it's better to use `JdbcLiteralFormatter`? Something along the lines of this when making the `check` a `StringBuilderSqlAppender`?

```suggestion
						final JdbcMapping jdbcMapping = (JdbcMapping) getDiscriminator().getType();
						jdbcMapping.getJdbcLiteralFormatter().appendJdbcLiteral( check, getDiscriminatorValue(), null );
```

Reply:
I think that doesn’t work because `getDiscriminatorValue()` returns a string.

Reply:
Ah, then you probably need to convert the string e.g.

```suggestion
						final JdbcMapping jdbcMapping = (JdbcMapping) getDiscriminator().getType();
						jdbcMapping.getJdbcLiteralFormatter().appendJdbcLiteral( check, jdbcMapping.getJavaType().fromString( getDiscriminatorValue() ), null );
```

Reply:
Is it necessary though, DYT?

If I'm not mistaken, we only support a _very_ limited range of types there, at least in annotations.

Reply:
Thanks, it does work.